pkg/ebpf/lockcontention.go

```diff
@@ -0,0 +1,492 @@
+	3: "hash-global-freelist-locks",
+	4: "percpu-lru-freelist-locks",
+	5: "lru-global-freelist-locks",
+	6: "lru-pcpu-freelist-locks",
+	7: "ringbuf-spinlock",
+	8: "ringbuf-waitq-spinlock",
+}
+
+// NewLockContentionCollector creates a prometheus.Collector for eBPF lock contention metrics
+func NewLockContentionCollector() *LockContentionCollector {
+	kversion, err := kernel.HostVersion()
+	if err != nil {
+		return nil
+	}
+
+	// the tracepoints for collecting lock contention information
+	// are only available after v5.19.0
+	// https://github.com/torvalds/linux/commit/16edd9b511a13e7760ed4b92ba4e39bacda5c86f
+	if kversion < kernel.VersionCode(5, 19, 0) {
+		return nil
+	}
+
+	ContentionCollector = &LockContentionCollector{
+		maxContention: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Subsystem: "ebpf__locks",
+				Name:      "_max",
+				Help:      "gauge tracking maximum time a tracked lock was contended for",
+			},
+			[]string{"name", "lock_type"},
+		),
+		avgContention: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Subsystem: "ebpf__locks",
+				Name:      "_avg",
+				Help:      "gauge tracking average time a tracked lock was contended for",
+			},
+			[]string{"name", "lock_type"},
+		),
+		totalContention: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Subsystem: "ebpf__locks",
+				Name:      "_total",
+				Help:      "counter tracking total time a tracked lock was contended for",
+			},
+			[]string{"name", "lock_type"},
+		),
+	}
+
+	return ContentionCollector
+}
+
+// Describe implements prometheus.Collector.Describe
+func (l *LockContentionCollector) Describe(descs chan<- *prometheus.Desc) {
+	l.maxContention.Describe(descs)
+	l.avgContention.Describe(descs)
+	l.totalContention.Describe(descs)
+}
+
+// Collect implements prometheus.Collector.Collect
+func (l *LockContentionCollector) Collect(metrics chan<- prometheus.Metric) {
+	l.mtx.Lock()
+	defer l.mtx.Unlock()
+
+	if !l.initialized {
+		return
+	}
+
+	var cursor ebpf.MapBatchCursor
+	lockRanges := make([]LockRange, l.ranges)
+	contention := make([]ContentionData, l.ranges)
+	if _, err := l.objects.LockStats.BatchLookup(&cursor, lockRanges, contention, nil); !errors.Is(err, ebpf.ErrKeyNotExist) {
+		log.Errorf("failed to perform batch lookup for lock stats: %v", err)
+		return
+	}
+
+	for i, data := range contention {
+		lr := lockRanges[i]
+		if lr.Start == 0 {
+			continue
+		}
+		mp, ok := l.trackedLockMemRanges[lr]
+		if !ok {
+			log.Errorf("found untracked lock range [0x%d, 0x%d+0x%d]", lr.Start, lr.Start, lr.Range)
+			continue
+		}
+
+		if (data.Total_time > 0) && (mp.totalTime != data.Total_time) {
+			avgTime := data.Total_time / uint64(data.Count)
+
+			lockType := lockTypes[lr.Type]
+			l.maxContention.WithLabelValues(mp.name, lockType).Set(float64(data.Max_time))
+			l.avgContention.WithLabelValues(mp.name, lockType).Set(float64(avgTime))
+
+			// TODO: should we consider overflows. u64 overflow seems very unlikely?
```

Comment:
I'd add a check that `mp.totalTime < data.Total_time` just in case. Overflow should never happen (`2^64` nanoseconds is ~600 years, I sure hope we don't have locks contending for that time) but apparently the `Add` panics if it receives a negative number, so I'd add the check just to avoid some invalid value of `Total_time` triggering a full panic.

Reply:
The reason I did not add this in the first place is because the overflow should never happen. If it does, then something else must have gone very wrong in which case a loud failure like a panic would be better to have.